File: java/05-reactor/src/test/java/com/fResult/reactor/ch5_04/EmitterProcessorTest.java
package com.fResult.reactor.ch5_04;

import lombok.val;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Sinks;
import reactor.test.StepVerifier;

class EmitterProcessorTest {
  @Test
  void emitterProcessor() {
    val processor = Sinks.many().multicast().<String>onBackpressureBuffer();
    produce(processor);
    consume(processor.asFlux());
  }

  private void produce(Sinks.Many<String> sink) {
    for (var i = 0; i < 3; i++) sink.tryEmitNext(String.valueOf(i + 1));
    sink.tryEmitComplete();
  }

  private void consume(Flux<String> publisher) {
    StepVerifier.create(publisher).expectNext("1").expectNext("2").expectNext("3").verifyComplete();
  }
}

File: kotlin/05-reactor/src/test/kotlin/com/fResult/reactor/ch5_04/EmitterProcessorTest.kt
package com.fResult.reactor.ch5_04

import org.junit.jupiter.api.Test
import reactor.core.publisher.Flux
import reactor.core.publisher.Sinks
import reactor.test.StepVerifier

class EmitterProcessorTest {
  @Test
  fun emitterProcessor() {
    val sinkEmitter = Sinks.many().multicast().onBackpressureBuffer<String>()
    produce(sinkEmitter)
    consume(sinkEmitter.asFlux())
  }

  private fun produce(sink: Sinks.Many<String>) {
    sink.tryEmitNext("1")
    sink.tryEmitNext("2")
    sink.tryEmitNext("3")
    sink.tryEmitComplete()
  }

  private fun consume(publisher: Flux<String>) {
    StepVerifier.create(publisher)
      .expectNext("1")
      .expectNext("2")
      .expectNext("3")
      .verifyComplete()
  }
}
